### PR TITLE
Use openapi3 for schema tests

### PR DESCRIPTION
**Pull request by hasufell**
# Issue Number

ADP-417

# Overview

- [x] added openapi3 as dep, fixed stack.cabal
- [x] fixed types and import, worked around lack of `servant-openapi` package
- [x] fix test failures
- [x] fix jormungandr

# Comments

### CODE REVIEW

File: stack.yaml
Comment (hasufell) [inline stack.yaml:44]: There's a difficult to track down build failure with `0.2.3` (which is the lowest bound for openapi3). It seems it only fails on a certain combination of packages, so I can't provide an upstream PR without further investigation.

File: lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs
Comment (hasufell) [inline lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs]: This is stolen from `Servant.Swagger.Test`.

File: specifications/api/swagger.yaml
Comment (hasufell) [inline specifications/api/swagger.yaml]: The oneOf had an overlap, because it doesn't check the format. So I think it's better to provide a lax javascript regex.

**Comment by hasufell** [comment]
This broke all of nix

**Comment by KtorZ** [comment]
@hasufell I've referenced the dependency (openapi3) from its github repository instead of its hackage package because the hackage index depends on which version of haskell.nix is used by the CI behind the scene. And the version we're currently using is too old so it doesn't know about openapi3.  

Bumping Haskell.nix is a possibility, but it sometimes can be quite some work which would be unnecessary _now_.

**Comment by rvl** [comment]
Thanks for fixing the nix @KtorZ .

@hasufell, for the record, here is [the documentation about updating the pins](https://github.com/input-output-hk/cardano-wallet/wiki/Building#haskellnix-pin), and I just added a link to the Haskell.nix ChangeLog file.

**Comment by hasufell** [comment]
> @hasufell, for the record, here is the documentation about updating the pins, and I just added a link to the Haskell.nix ChangeLog file.

I read that already and it didn't work:

```
$ niv update haskell.nix
Update haskell.nix
  INFO: new sources.nix available: 18 -> 23
    Please run 'niv init' or add the following line in the nix/sources.nix file:
    # niv: no_update
Done: Update haskell.nix
$ nix-shell
unpacking 'https://github.com/input-output-hk/haskell.nix/archive/23539aef9ea3591dc0d218c5f60b2a4f7d229a20.tar.gz'...
unpacking 'https://github.com/input-output-hk/hackage.nix/archive/7f67b08b9c6f5e9bd5128bf06de54f5de012ee60.tar.gz'...
warning: dumping very large path (> 256 MiB); this may run out of memory
trace: Using IOHK default nixpkgs
trace: gitSource.nix: /nix/store/dvp437wx7bmj4kfpvybifnj0aak27hmc-cardano-node-src does not seem to be a git repository,
assuming it is a clean checkout.
trace: Using index-state: 2020-07-15T00:00:00Z for cardano-node-src
trace: Using index-state: 2020-07-15T00:00:00Z for cardano-node-src
trace: Cleaning component source not supported for hpack package: cardano-addresses-2.1.0
error: The Haskell package set does not contain the package: indexed-profunctors (build dependency).
If you are using Stackage, make sure that you are using a snapshot that contains the package. Otherwise you may need to update the Hackage snapshot you are using, usually by updating haskell.nix.
```

**Comment by rvl** [comment]
```
error: The Haskell package set does not contain the package: indexed-profunctors (build dependency).
If you are using Stackage, make sure that you are using a snapshot that contains the package.
```

What this means is that the Stackage snapshot doesn't contain indexed-profunctors. So it needed the change that @KtorZ added.

The Haskell.nix rev is now updated on master anyway. I have pushed a fix.

**Comment by hasufell** [comment]
> What this means is that the Stackage snapshot doesn't contain indexed-profunctors. So it needed the change that @KtorZ added.

Can we add documentation about this? What are the options when this case happens and how to execute the different solutions.

**Comment by rvl** [comment]
Sure. But note that the nix build is based on `stack.yaml`. So if the stack build fails due to dependencies not in stackage, then the nix build is also going to fail.

**Comment by hasufell** [comment]
@rvl please don't do reverse merges back from master. It messes up the history and makes rebasing etc harder.

**Comment by rvl** [comment]
What is the actual problem @hasufell? It should be easy for you to rebase, and then the merge commit disappears.
I wasn't going to rebase and force push your branch out of consideration for you, and also because it sometimes happens that the other person just force pushes the branch back to what it was before, and the changes are lost.

**Comment by hasufell** [comment]
>  It should be easy for you to rebase, and then the merge commit disappears.

That's only true if you rebase against upstream master and that now causes lots of other conflicts. Squashing isn't easily possible anymore either (which reduces merge conflicts during rebases against upstream branches).

**Comment by rvl** [comment]
It's easy. There are two merge conflicts which you would have needed to resolve anyway before merging into master.

**Comment by hasufell** [comment]
> It's easy. There are two merge conflicts which you would have needed to resolve anyway before merging into master.

I'd prefer not to have reverse merges in my branches. Thanks :)

File: lib/test-utils/src/Test/Utils/OpenApi.hs
Comment (hasufell) [inline lib/test-utils/src/Test/Utils/OpenApi.hs]: This is a copy-paste. We might want to fork servant-swagger, but that will be more work.

File: lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs
Comment (hasufell) [inline lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:1032]: https://github.com/input-output-hk/cardano-addresses/blob/350abfe36c3fc2949da567ca47a957a9d5789c92/core/lib/Cardano/Address/Script/Parser.hs#L61-L88

File: lib/core/cardano-wallet-core.cabal
Comment (hasufell) [inline lib/core/cardano-wallet-core.cabal:235]: I don't know if it builds on windows, but it has the code bundled.

File: cabal.project
Comment (KtorZ) [inline cabal.project]: release is coming soon.

File: lib/core/cardano-wallet-core.cabal
Comment (KtorZ) [reply lib/core/cardano-wallet-core.cabal:235]: `bors try` will tell us this. There's a hydra job that builds the wallet for Windows, from Linux though.

File: lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:1949]: :+1:

File: specifications/api/swagger.yaml
Comment (KtorZ) [inline specifications/api/swagger.yaml:311]: :ok_hand:

**Comment by KtorZ** [comment]
LGTM, modulo the revision on cardano-addresses which we need to address (release for this one is coming today hopefully).

**Comment by hasufell** [comment]
bors try

**Review by KtorZ — APPROVED** [review]

**Pull request merged**
